py/redrock/rebin.py

```diff
@@ -125,24 +239,150 @@ def trapz_rebin(x, y, xnew=None, edges=None):
 
     return result
 
+def trapz_rebin_batch_gpu(x, y, xnew=None, edges=None, myz=None):
+    """Rebin y(x) flux density using trapezoidal integration between bin edges
+    GPU algorithm can rebin in batch for multiple redshifts and bases and
+    returns 1-d, 2-d, or 3-d array (n redshifts x n bins x n basis) where
+    n basis is the optional second dimension of the y input array and
+    n redshifts is the length of the optional myz array.
+
+    Notes:
+        y is interpreted as a density, as is the output, e.g.
+
+        >>> x = np.arange(10)
+        >>> y = np.ones(10)
+        >>> trapz_rebin(x, y, edges=[0,2,4,6,8])  #- density still 1, not 2
+        array([ 1.,  1.,  1.,  1.])
+        >>> y = np.ones((2,10)) #nbasis = 2
+        >>> y[1,:] = np.arange(10)
+        >>> rebin.trapz_rebin(x, y, edges=[0,2,4,6,8], use_gpu=True)
+        array([[1., 1.],
+               [1., 3.],
+               [1., 5.],
+               [1., 7.]])
+        >>> rebin.trapz_rebin(x, y, edges=[0,2,4,6,8], myz=[0,0.5],use_gpu=True)
+        array([[[1. , 1. ],
+                [1. , 3. ],
+                [1. , 5. ],
+                [1. , 7. ]],
+               [[1. , 0.5],
+                [1. , 1.5],
+                [1. , 2.5],
+                [1. , 3.5]]])
+
+
+    Args:
+        x (1-d array): input x values.
+        y (1-d or 2-d array): input y values (for all bases).
+        edges (1-d array): (optional) new bin edges.
+        myz (array): (optional) redshift array to rebin in batch, applying
+            redshifts on-the-fly to x
+
+
+    Returns:
+        cp.array: integrated results with shape (nz, nbin, nbasis)
+            where nbin = len(results) = len(edges)-1
+            if nz or nbasis is 1, this dimension will be squeezed,
+            e.g. for nbasis = 1 and nz = 100, the shape will be (100, nbin)
+            while for nbasis = 1 and nz = 1 the shape will be (nbin).
+
+    Raises:
+        ValueError: if edges are outside the range of x or if len(x) != len(y)
+
+    """
+    #import cupy here
+    import cupy as cp
+
+    if edges is None:
+        edges = centers2edges(xnew)
+    else:
+        edges = np.asarray(edges)
+    edges = cp.array(edges, dtype=cp.float64)
 
-def rebin_template(template, z, dwave):
+    if (myz is None):
+        myz = cp.array([0], dtype=cp.float64)
+    myz = cp.array(myz, dtype=cp.float64)
+
+    x = cp.array(x, dtype=cp.float64)
+    y = cp.array(y, dtype=cp.float64)
```

Comment:
Let's use `cp.asarray` to avoid copying the arrays if `x`, `y`, `edges`, and/or `myz` are already cupy arrays.

Reply:
Done - cp.array() has been updated to cp.asarray() in all cases in rebin.py and utils.py